ocaml/lambda/translprim.ml

```diff
@@ -488,79 +488,86 @@ let simplify_constant_constructor = function
+                                      |
+      /------+---------+--------------+--------------+---------------\
+      |      |         |              |              |               |
+    addr  float  unboxed-float  unboxed-int32  unboxed-int64  unboxed-nativeint
+      |
+    int
    Note that the GLB is not guaranteed to exist, in which case we return
    our first argument instead of raising a fatal error because, although
    it cannot happen in a well-typed program, (ab)use of Obj.magic can
    probably trigger it.
 *)
 let glb_array_type t1 t2 =
   match t1, t2 with
-  (* No GLB; only used in the [Obj.magic] case *)
-  | Pfloatarray, (Paddrarray | Pintarray)
-  | (Paddrarray | Pintarray), Pfloatarray -> t1
-
-  (* Compute the correct GLB *)
   | Pgenarray, x | x, Pgenarray -> x
   | Paddrarray, x | x, Paddrarray -> x
   | Pintarray, Pintarray -> Pintarray
   | Pfloatarray, Pfloatarray -> Pfloatarray
+  | Punboxedfloatarray, Punboxedfloatarray -> Punboxedfloatarray
+  | Punboxedintarray Pint32, Punboxedintarray Pint32 -> Punboxedintarray Pint32
+  | Punboxedintarray Pint64, Punboxedintarray Pint64 -> Punboxedintarray Pint64
+  | Punboxedintarray Pnativeint, Punboxedintarray Pnativeint ->
+    Punboxedintarray Pnativeint
+  (* No GLB; only used in the [Obj.magic] case *)
+  | (Pfloatarray | Pintarray | Punboxedfloatarray | Punboxedintarray _),
+    (Pfloatarray | Pintarray | Punboxedfloatarray | Punboxedintarray _) -> t1
 
 let glb_array_ref_type t1 t2 =
   match t1, t2 with
-  (* No GLB; only used in the [Obj.magic] case *)
-  | Pfloatarray_ref _, (Paddrarray | Pintarray)
-  | (Paddrarray_ref | Pintarray_ref), Pfloatarray -> t1
-
-  (* Compute the correct GLB *)
-
-  (* Pgenarray >= _ *)
+  (* Pgenarray >= _ (for value layouts) *)
   | (Pgenarray_ref _ as x), Pgenarray -> x
   | Pgenarray_ref _, Pintarray -> Pintarray_ref
   | Pgenarray_ref _, Paddrarray -> Paddrarray_ref
   | Pgenarray_ref mode, Pfloatarray -> Pfloatarray_ref mode
   | (Paddrarray_ref | Pintarray_ref | Pfloatarray_ref _ as x), Pgenarray -> x
-
   (* Paddrarray > Pintarray *)
   | Paddrarray_ref, Paddrarray -> Paddrarray_ref
   | Paddrarray_ref, Pintarray -> Pintarray_ref
   | Pintarray_ref, Paddrarray -> Pintarray_ref
-
   (* Pintarray is a minimum *)
   | Pintarray_ref, Pintarray -> Pintarray_ref
-
   (* Pfloatarray is a minimum *)
   | (Pfloatarray_ref _ as x), Pfloatarray -> x
+  (* Likewise for the unboxed array kinds *)
+  | Punboxedfloatarray_ref, Punboxedfloatarray -> t1
+  | Punboxedintarray_ref Pint32, Punboxedintarray Pint32 -> t1
+  | Punboxedintarray_ref Pint64, Punboxedintarray Pint64 -> t1
+  | Punboxedintarray_ref Pnativeint, Punboxedintarray Pnativeint -> t1
+  (* No GLB; only used in the [Obj.magic] case *)
+  | (Pgenarray_ref _ | Paddrarray_ref | Pfloatarray_ref _ | Pintarray_ref
+      | Punboxedfloatarray_ref | Punboxedintarray_ref _),
+    (Pgenarray | Pfloatarray | Pintarray | Paddrarray | Punboxedfloatarray
+      | Punboxedintarray _) -> t1
```

Comment:
The unboxed cases here are probably only reachable via something like:

```ocaml
external get : float# array -> int -> float = "%floatarray_safe_get"
let d = get [|#1.|] 0
```

Should we consider erroring here for layout mismatch or is this a desirable behavior to keep around?

Alternative would be to do something like:

```ocaml
  | Punboxedfloatarray_ref, Punboxedfloatarray -> Punboxedfloatarray_ref
  | Punboxedfloatarray_ref, _ | _, Punboxedfloatarray ->
    Misc.fatal_error "..."

  | Punboxedintarray_ref Pnativeint, Punboxedintarray Pnativeint ->
    Punboxedintarray_ref Pnativeint
  | Punboxedintarray_ref Pint32, Punboxedintarray Pint32 -> Punboxedintarray_ref Pint32
  | Punboxedintarray_ref Pint64, Punboxedintarray Pint64 -> Punboxedintarray_ref Pint64
  | Punboxedintarray_ref _, _ | _, Punboxedintarray _ ->
    Misc.fatal_error "..."
```

Reply:
Probably best to discuss this with @goldfirere and @ccasin

Reply:
Talked with @goldfirere about this. The takeway from the conversation is a bit more in favor of throwing errors for the unboxed cases. The main arguments being:

1. Mixing the different layouts doesn't feel quite right
2. It would be rare to run into these errors: you will have to declare externals that call array primitives with the wrong kind (more likely to be errors than intentional?)
3. Doesn't prevent someone from achieving the same behavior if they really want to. With the use of `Obj.magic`, something like `Floatarray.get (Obj.magic float_u_array : floatarray) 0` would still work and not hit cases in question

I don't feel strongly about making this change and the current implementation would also be fine.

Reply:
I'm also in favour of more errors here.  @alanechang could you please implement what you discussed?  (Make a PR against this one on my repo)